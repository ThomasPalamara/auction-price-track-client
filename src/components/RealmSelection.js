import React from 'react';
import Select from 'react-select';
import { Button } from 'antd';
import config from '../config';

export default class realmSelection extends React.Component {
  state = {
    status: undefined,
    error: undefined,
    selectedOption: {},
    response: [],
    realm: '',
  };

  componentDidMount() {
    fetch(`${config.serverUrl}/api/realms`)
    .then(res => res.json())
    .then(response => this.setState({ response: response}));
  }

  handleRealmPicked = e => {
    e.preventDefault();
    if(this.state.selectedOption && this.state.response.find(x => x.label === this.state.selectedOption.label) ){
    const realm = this.state.selectedOption.label;
      this.setState({status: 'done'})
      this.props.handleRealmPicked(realm)
    }else{
      this.setState({selectedOption: null})
      this.setState({error: 'Veuillez selectionner un nom de royaume valide'})
    }

    
  }

  handleChange = (selectedOption) => {
    this.setState({ selectedOption });
    console.log(`Option selected:`, selectedOption);
  }

  render() {
    return (
      <div className="realmSelection">
      <h2 className={`step ${this.state.status}`}>1/ Choisissez votre royaume</h2>
        <form onSubmit={this.handleRealmPicked}>
        <div className="select-realm">
        <Select
          value={this.state.selectedOption}
          onChange={this.handleChange}
          name="realm"
          options={this.state.response}
        />
        </div>
        <Button id="chooseRealmBtn" htmlType="submit" className="btn btn-large" loading={this.props.loading} onClick={this.enterLoading}>
          Go !
        </Button>
        </form>
      </div>
    );
  }
}

          //<Select options={this.state.response.map(e => e.name)}/>

          // <AutoComplete
          //   items = {this.state.response}
          //   shouldItemRender={(item, value) => item.name.toLowerCase().indexOf(value.toLowerCase()) > -1}
          //   getItemValue={item => item.name}
          //   renderItem={(item, highlighted) =>
          //     <div
          //       key={item.slug}
          //       style={{ backgroundColor: highlighted ? '#eee' : 'transparent', textAlign: 'left', paddingLeft: '5px', fontSize: '15px'}}
          //     >
          //       {item.name}
          //     </div>
          //   }
          //   value={this.state.value}
          //   onChange={e => this.setState({ value: e.target.value })}
          //   onSelect={value => this.setState({ value })}
          // />