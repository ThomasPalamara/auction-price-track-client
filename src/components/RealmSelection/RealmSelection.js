import React, { useEffect, useState } from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';
import { Divider } from 'antd';
import { apiURL } from '../../constants';

const RealmSelection = (props) => {

  const [status, setStatus] = useState(undefined);
  const [error, setError] = useState(undefined);
  const [selectedOption, setSelectedOption] = useState({ value: '', label: '' });
  const [realms, setRealms] = useState([]);

  useEffect(() => {
    fetch(`${apiURL}/realms`)
      .then((res) => {
        if (!res.ok) {
          throw new Error('Not found');
        }
        return res.json();
      })
      .then((response) => {
        setRealms(response.map(realm => ({ value: realm.slug, label: realm.name })));
      });
  }, []);

  useEffect(() => {
    const { handleRealmPicked } = props;

    if (selectedOption && realms.find(x => (x.value === selectedOption.value))) {
      const realm = selectedOption;
      setError('');
      handleRealmPicked(realm);
    } else {
      setSelectedOption(null);
      setError('Veuillez selectionner un nom de royaume valide');
    }
  }, [selectedOption]);

  const handleChange = (selected) => {
    setSelectedOption({ value: selected.value, label: selected.label });
  };

  return (
    <div className="realmSelection">
      <Divider orientation="left">
        <h3 className={`step ${status}`}>1/ Choisissez votre royaume</h3>
      </Divider>
      <div className="select-realm">
        <Select
          value={selectedOption}
          onChange={handleChange}
          name="realm"
          isClearable
          isSearchable
          options={realms.map(realm => realm)}
        />
        {error && <span className="error">{error}</span>}
      </div>
    </div>
  );
}

RealmSelection.propTypes = { handleRealmPicked: PropTypes.func.isRequired };

export default RealmSelection;
