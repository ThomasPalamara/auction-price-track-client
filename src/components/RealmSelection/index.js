import RealmSelection from './RealmSelection';

export default RealmSelection;
