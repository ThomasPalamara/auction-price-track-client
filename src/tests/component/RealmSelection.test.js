import React from 'react';
import { shallow } from "enzyme";
import toJSON from 'enzyme-to-json';
import RealmSelection from '../../components/RealmSelection';
import realms from '../fixtures/realms'

test('Should render RealmSelection component correctly', () => {
    const wrapper = shallow(<RealmSelection />);
    wrapper.setState({ response: realms })
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should render error correctly for empty input', () => {
    const wrapper = shallow(<RealmSelection />);
    wrapper.setState({ response: realms })

    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error')).toBeTruthy()
    expect(wrapper.find('label.error').length).toBe(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should render error correctly for invalid input', () => {
    const wrapper = shallow(<RealmSelection handleRealmPicked={() => {}}/>);
    wrapper.setState({ response: realms });

    const select = wrapper.find('StateManager').at(0).simulate('change', {
        "label": "Ysondres",
        "value": "ysondres",
      });
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error')).toBeTruthy();
    expect(wrapper.find('label.error').length).toBe(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should set selected realm correctly', () => {
    const wrapper = shallow(<RealmSelection handleRealmPicked={() => {}}/>);
    wrapper.setState({ response: realms });
    const selectedOption = {
        "label": "Ysondres",
        "value": "ysondres",
      }

    const select = wrapper.find('StateManager').at(0).simulate('change', selectedOption);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('selectedOption')).toEqual(selectedOption);
});

// test('Should handle realm selection ', () => {
//     const wrapper = shallow(<RealmSelection />);
//     const select = wrapper.find('StateManager').at(0);
//     select.value = 'Ysondres'
//     wrapper.find('form').simulate('submit', {
//         preventDefault: () => { }
//     });
// })