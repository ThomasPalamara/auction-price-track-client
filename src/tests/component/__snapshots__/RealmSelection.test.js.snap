// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render RealmSelection component correctly 1`] = `
<div
  className="realmSelection"
>
  <Divider
    orientation="left"
  >
     
    <h3
      className="step undefined"
    >
      1/ Choisissez votre royaume
    </h3>
     
  </Divider>
  <form
    onSubmit={[Function]}
  >
    <div
      className="select-realm"
    >
      <StateManager
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        name="realm"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Elune",
              "value": "elune",
            },
            Object {
              "label": "Ysondre",
              "value": "ysondre",
            },
            Object {
              "label": "Krasus",
              "value": "krasus",
            },
          ]
        }
        value={Object {}}
      />
    </div>
    <Button
      block={false}
      className="btn btn-large"
      ghost={false}
      htmlType="submit"
      id="chooseRealmBtn"
      loading={false}
      prefixCls="ant-btn"
    >
      Go !
    </Button>
  </form>
</div>
`;

exports[`Should render error correctly for empty input 1`] = `
<div
  className="realmSelection"
>
  <Divider
    orientation="left"
  >
     
    <h3
      className="step undefined"
    >
      1/ Choisissez votre royaume
    </h3>
     
  </Divider>
  <form
    onSubmit={[Function]}
  >
    <div
      className="select-realm"
    >
      <StateManager
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        name="realm"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Elune",
              "value": "elune",
            },
            Object {
              "label": "Ysondre",
              "value": "ysondre",
            },
            Object {
              "label": "Krasus",
              "value": "krasus",
            },
          ]
        }
        value={null}
      />
      <label
        className="error"
      >
        Veuillez selectionner un nom de royaume valide
      </label>
    </div>
    <Button
      block={false}
      className="btn btn-large"
      ghost={false}
      htmlType="submit"
      id="chooseRealmBtn"
      loading={false}
      prefixCls="ant-btn"
    >
      Go !
    </Button>
  </form>
</div>
`;

exports[`Should render error correctly for invalid input 1`] = `
<div
  className="realmSelection"
>
  <Divider
    orientation="left"
  >
     
    <h3
      className="step undefined"
    >
      1/ Choisissez votre royaume
    </h3>
     
  </Divider>
  <form
    onSubmit={[Function]}
  >
    <div
      className="select-realm"
    >
      <StateManager
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultValue={null}
        name="realm"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Elune",
              "value": "elune",
            },
            Object {
              "label": "Ysondre",
              "value": "ysondre",
            },
            Object {
              "label": "Krasus",
              "value": "krasus",
            },
          ]
        }
        value={null}
      />
      <label
        className="error"
      >
        Veuillez selectionner un nom de royaume valide
      </label>
    </div>
    <Button
      block={false}
      className="btn btn-large"
      ghost={false}
      htmlType="submit"
      id="chooseRealmBtn"
      loading={false}
      prefixCls="ant-btn"
    >
      Go !
    </Button>
  </form>
</div>
`;
